packages/flutter/lib/src/material/material.dart

```diff
@@ -617,27 +628,37 @@ abstract class InkFeature {
   /// Called when the ink feature is no longer visible on the material.
   final VoidCallback? onRemoved;
 
-  bool _debugDisposed = false;
+  bool _disposed = false;
+
+  /// Whether or not visual reaction is activated.
+  ///
+  /// Change this field will affect whether this InkFeature is render in next
+  /// frame.
+  ///
+  /// For this InkFeature to render properly, it should usually be change in
+  /// [State.deactivate] and [State.activate].
+  bool visible = true;
```

Comment:
When this value is changed, how is the controller told that it needs to redraw to show/hide this feature?

Reply:
Now that I think about it, it's not a good name. It is not to be used as a general visibility control, but used exclusively by `referenceBox.activate` and `deactivate`. This is needed because there will always be a `paint` between `referenceBox.deactivate` and `dispose`:

```
Painting InkSplash#89803 visible true
Painting InkSplash#89803 visible true
Painting InkSplash#89803 visible true
Painting InkSplash#89803 visible true
Setting InkSplash#89803 visible false
Painting InkSplash#89803 visible false
Disposing InkSplash#89803
Painting InkSplash#ab32a visible true
Painting InkSplash#ab32a visible true
Painting InkSplash#ab32a visible true
Painting InkSplash#ab32a visible true
Setting InkSplash#ab32a visible false
Painting InkSplash#ab32a visible false
Disposing InkSplash#ab32a
Painting InkSplash#751e1 visible true
Painting InkSplash#751e1 visible true
Painting InkSplash#751e1 visible true
Painting InkSplash#751e1 visible true
Setting InkSplash#751e1 visible false
Setting InkSplash#751e1 visible true
Painting InkSplash#751e1 visible true
Painting InkSplash#751e1 visible true
Painting InkSplash#751e1 visible true
Painting InkSplash#26191 visible true
Painting InkHighlight#414cf visible true
Painting InkSplash#26191 visible true
Painting InkHighlight#414cf visible true
Painting InkSplash#26191 visible true
Painting InkHighlight#414cf visible true
Painting InkSplash#26191 visible true
Painting InkHighlight#414cf visible true
Setting InkSplash#26191 visible false
Setting InkHighlight#414cf visible false
Setting InkSplash#26191 visible true
Setting InkHighlight#414cf visible true
Painting InkSplash#26191 visible true
Painting InkHighlight#414cf visible true
Painting InkSplash#26191 visible true
Painting InkHighlight#414cf visible true
Painting InkSplash#26191 visible true
Painting InkHighlight#414cf visible true
Painting InkSplash#26191 visible true
Painting InkSplash#26191 visible true
```